### PR TITLE
Added problem and solution for cf-1713E

### PR DESCRIPTION
**Pull request by Keys9873**
_If any of the below don't apply to this Pull Request, mark the checkbox as done._

- [_] I have tested my code.
- [_] I have added my solution according to the steps [here](https://usaco.guide/general/adding-solution#steps).
- [_] I have followed the code conventions mentioned [here](https://usaco.guide/general/adding-solution/#code-conventions), which include the following: 
  - I understand that if it is clear that I have not attempted to follow these guidelines (ex. if I have not used tabs to indent), my PR will be closed.
  - If changes are requested, I will re-request the review after making the changes.

### CODE REVIEW

**Comment by SansPapyrus683** [comment]
bro
i said notify me before adding a new problem

**Comment by Keys9873** [comment]
Oh was it on discord? I'm not really active ...

File: solutions/cf-1713E.mdx [hunk 1]
```diff
@@ -0,0 +1,134 @@
...
+
+## Solution
+
+[Official Analysis](https://codeforces.com/blog/entry/105419)
+
+Notice that $a[i][j]$ can only be swapped with $a[j][i]$ no matter which operations are performed.
+
+In addition, we want to lexicographically sort the array. Thus, we can iterate and greedily process the array.
+
+For all $j$ such that $i < j$ (we only need to consider each pair once), we can determine whether we want to swap them or keep them the same.
```
Comment (Aadit-Ambadkar) [inline solutions/cf-1713E.mdx:16]:
```suggestion
For all $j$ such that $i < j$ (we only need to consider each pair once and keeping $i < j$ will ensure this), we can determine whether we want to swap them or keep them the same.
```

File: solutions/cf-1713E.mdx [hunk 2]
```diff
@@ -0,0 +1,134 @@
...
+using vi = vector<int>;
+
+struct DSU {
+	vi e;
+	DSU(int n) {
+		e = vi(n, -1);
+	}
+	int get(int x) {
+		return e[x] < 0 ? x : e[x] = get(e[x]);
+	}
```
Comment (Aadit-Ambadkar) [inline solutions/cf-1713E.mdx:52]:
```suggestion
	int get(int x) {
		if (e[x] < 0) return x;
		else {
			e[x] = get(e[x]);
			return e[x];
		}
	}
```

File: solutions/cf-1713E.mdx [hunk 3]
```diff
@@ -0,0 +1,134 @@
...
+	for (int i = 0; i < n; i++)
+	  for (int j = 0; j < n; j++) cin >> a[i][j];
+	DSU dsu(2 * n);
+	for (int i = 0; i < n; i++) {
+		for (int j = i + 1; j < n; j++) { // Only check each pair once and diagnals don't matter
+			// If they are the same we should not merge either way because it could affect later merges
+			if (a[i][j] == a[j][i]) continue;
+			else if (a[i][j] > a[j][i]) {
+				// Try to swap the elements
+				if (dsu.get(i) != dsu.get(j)) { // Check if the states are already merged the opposite way
```
Comment (Aadit-Ambadkar) [inline solutions/cf-1713E.mdx:81]: LINE LENGTH EXCEEDS 80 CHARACTERS

File: solutions/cf-1713E.mdx [hunk 4]
```diff
@@ -0,0 +1,134 @@
...
+					ans[i][j] = a[i][j];
+					ans[j][i] = a[j][i];
+				} else {
+					ans[i][j] = a[j][i];
+					ans[j][i] = a[i][j];
+				}
+			}
+		}
+	}
+  // Print the answer
```
Comment (Aadit-Ambadkar) [inline solutions/cf-1713E.mdx:114]:
```suggestion
	// Print the answer
```

**Review by Aadit-Ambadkar — CHANGES_REQUESTED** [review]

File: solutions/cf-1713E.mdx (hunk 3 above)
Comment (Keys9873) [reply solutions/cf-1713E.mdx]: fixed

File: solutions/cf-1713E.mdx [hunk 5]
```diff
@@ -0,0 +1,145 @@
...
+		return true;
+	}
+};
+
+void solve() {
+	int n;
+	cin >> n;
+	int a[n][n];
+	for (int i = 0; i < n; i++)
+	  for (int j = 0; j < n; j++) cin >> a[i][j];
```
Comment (SansPapyrus683) [inline solutions/cf-1713E.mdx:81]: sus

File: solutions/cf-1713E.mdx [hunk 6]
```diff
@@ -0,0 +1,145 @@
...
+
+In addition, we want to lexicographically sort the array. Thus, we can iterate and greedily process the array.
+
+For all $j$ such that $i < j$ (we only need to consider each pair once and keeping $i < j$ will ensure this), we can determine whether we want to swap them or keep them the same.
+
+If $a[i][j] > a[j][i]$, we want to swap them. If $a[i][j] < a[j][i]$ we want to keep them the same. If they are equal it doesn't matter.
+
+However, it is possible that not all conditions can work together with each other and swapping something means we can't swap something else later on.
+
+We can use the [Amogus Trick](https://codeforces.com/blog/entry/106407) to represent this problem. Create a disjoint set union of $2n$ nodes where node $i$ represents the state where the operation is performed on $i$ and $i + n$ represents the state where operation is not performed on $i$.
```
Comment (SansPapyrus683) [inline solutions/cf-1713E.mdx:22]: is there no other way to solve this problem?
because if the only way to solve this is with some meme-a** trick, i don't think this is a good problem

Comment (SansPapyrus683) [reply solutions/cf-1713E.mdx:22]: or unless you can give a nice intuitive method for arriving at this trick

File: solutions/cf-1713E.mdx [hunk 7]
```diff
@@ -0,0 +1,145 @@
...
+				// Keep the elements the same if the states
+				// are not already merged the opposite way
+				if (dsu.get(i) != dsu.get(j + n)) {
+					dsu.unite(i, j);
+					dsu.unite(i + n, j + n);
+				}
+			}
+		}
+	}
+	// Construct the answer
```
Comment (SansPapyrus683) [inline solutions/cf-1713E.mdx:105]: separate relevant code segments w/ whitespace

**Review by SansPapyrus683 — CHANGES_REQUESTED** [review]

File: solutions/cf-1713E.mdx (hunk 5 above)
Comment (SansPapyrus683) [reply solutions/cf-1713E.mdx:81]: braces + tabs

File: solutions/cf-1713E.mdx (hunk 6 above)
Comment (SansPapyrus683) [reply solutions/cf-1713E.mdx:22]: i wouldn't mention the trick then if it isn't really needed

File: solutions/cf-1713E.mdx (hunk 5 above)
Comment (Keys9873) [reply solutions/cf-1713E.mdx]: added

File: solutions/cf-1713E.mdx (hunk 6 above)
Comment (Keys9873) [reply solutions/cf-1713E.mdx]: fixed

File: solutions/cf-1713E.mdx (hunk 7 above)
Comment (Keys9873) [reply solutions/cf-1713E.mdx:111]: fixed

File: solutions/cf-1713E.mdx [hunk 8]
```diff
@@ -0,0 +1,152 @@
...
+	for (int i = 0; i < n; i++) {
+		for (int j = 0; j < n; j++) {
+			cout << ans[i][j] << " ";
+		}
+		cout << endl;
+	}
+}
+
+int main() { 
+	ios::sync_with_stdio(false); cin.tie(nullptr);
```
[the remark this answers is not in the document]
Comment (Keys9873) [reply solutions/cf-1713E.mdx:142]: Yes

[the remark this answers is not in the document]
Comment (Keys9873) [reply solutions/cf-1713E.mdx:142]: Even if it wasn't necessary, shouldn't you put fast io regardless?

File: solutions/cf-1713E.mdx [hunk 9]
```diff
@@ -0,0 +1,152 @@
...
+	
+	// Construct the answer
+	int ans[n][n];
+	for (int i = 0; i < n; i++) {
+		for (int j = i; j < n; j++) {
+			/* If a[i][j] is part of a diagonal or the pair is equal (a[i][j] == a[j][i]),
+			 * manually set it to be the same as the original array
+			 */
+			if (i == j) ans[i][j] = a[i][j];
+			else if (a[i][j] == a[j][i]) ans[i][j] = ans[j][i] = a[i][j];
```
[the remark this answers is not in the document]
Comment (Keys9873) [reply solutions/cf-1713E.mdx]: fixed

File: solutions/cf-1713E.mdx (hunk 8 above)
[the remark this answers is not in the document]
Comment (SansPapyrus683) [reply solutions/cf-1713E.mdx:145]: ? no?

File: solutions/cf-1713E.mdx [hunk 10]
```diff
@@ -0,0 +1,155 @@
...
+using vi = vector<int>;
+
+struct DSU {
+	vi e;
+	DSU(int n) {
+		e = vi(n, -1);
+	}
+	
+	int get(int x) {
+		if (e[x] < 0) return x;
```
Comment (SansPapyrus683) [inline solutions/cf-1713E.mdx:52]: use braces _everywhere_

File: solutions/cf-1713E.mdx [hunk 11]
```diff
@@ -0,0 +1,155 @@
...
+
+<CPPSection>
+
+```cpp
+#include <bits/stdc++.h>
+using namespace std;
+
+using vi = vector<int>;
+
+struct DSU {
```
Comment (SansPapyrus683) [inline solutions/cf-1713E.mdx:45]: also put this in a codesnip or smth

File: solutions/cf-1713E.mdx (hunk 10 above)
Comment (Keys9873) [reply solutions/cf-1713E.mdx]: fixed

File: solutions/cf-1713E.mdx [hunk 12]
```diff
@@ -42,6 +42,7 @@ using namespace std;
 
 using vi = vector<int>;
 
+//BeginCodeSnip{}
```
Comment (SansPapyrus683) [inline solutions/cf-1713E.mdx:45]:
```suggestion
//BeginCodeSnip{DSU}
```

File: solutions/cf-1713E.mdx (hunk 11 above)
Comment (Keys9873) [reply solutions/cf-1713E.mdx:46]: done

File: solutions/cf-1713E.mdx [hunk 13]
```diff
@@ -0,0 +1,164 @@
...
+		}
+	}
+	
+	// Construct the answer
+	int ans[n][n];
+	for (int i = 0; i < n; i++) {
+		for (int j = i; j < n; j++) {
+			/* If a[i][j] is part of a diagonal or the pair is equal (a[i][j] == a[j][i]),
+			 * manually set it to be the same as the original array
+			 */
```
Comment (SansPapyrus683) [inline solutions/cf-1713E.mdx:127]:
```suggestion
			/*
			 * If a[i][j] is part of a diagonal or the pair is equal (a[i][j] == a[j][i]),
			 * manually set it to be the same as the original array
			 */
```

File: solutions/cf-1713E.mdx [hunk 14]
```diff
@@ -0,0 +1,164 @@
...
+					ans[j][i] = a[j][i];
+				} else {
+					ans[i][j] = a[j][i];
+					ans[j][i] = a[i][j];
+				}
+			}
+		}
+	}
+	
+	// Print the answer
```
Comment (SansPapyrus683) [inline solutions/cf-1713E.mdx:144]:
```suggestion
```
comments need not state the obvious

File: solutions/cf-1713E.mdx (hunk 8 above)
[the remark this answers is not in the document]
Comment (SansPapyrus683) [reply solutions/cf-1713E.mdx:154]: bump

File: solutions/cf-1713E.mdx [hunk 15]
```diff
@@ -0,0 +1,164 @@
...
+			cout << ans[i][j] << " ";
+		}
+		cout << endl;
+	}
+}
+
+int main() { 
+	ios::sync_with_stdio(false); cin.tie(nullptr);
+	int T = 1; cin >> T;
+	while (T--) {
```
Comment (SansPapyrus683) [inline solutions/cf-1713E.mdx:156]:
```suggestion
	int T;
	cin >> T;
	for (int t = 0; t < T; t++) {
```

**Review by SansPapyrus683 — CHANGES_REQUESTED** [review]

File: solutions/cf-1713E.mdx (hunk 8 above)
[the remark this answers is not in the document]
Comment (Keys9873) [reply solutions/cf-1713E.mdx:154]: Yes you need it
I tried without and it tle'd

File: solutions/cf-1713E.mdx [hunk 16]
```diff
@@ -0,0 +1,165 @@
...
+
+We can create a disjoint set union to model this problem. The DSU will have $2n$ nodes where node $i$ represents the state where the operation is performed on the $i$th row and node $i + n$ represents the state where operation is not performed on the $i$th row.
+
+As we are iterating greedily through the array, we can merge ($i$ and $j$) and ($i + n$ and $j + n$) if we don't want to swap. This works because if we swap neither or swap both the arrangement remains the same.
+If we do want to swap we can merge ($i$ and $j + n$) and ($i + n$ and $j$). This works because if we swap exactly one of these, then the arrangement will change. 
+
+After this we can construct the answer.
+
+We can iterate over the entire grid. If it is a diaganal it will never change. If $a[i][j] == a[j][i]$ then we can manually set them to normal (since we didn't merge them in the disjoint set union when comparing them the states might have not been merged). If it is not a diaganol we can check whether $i$ and $j$ are merged in the disjoint set union. If they aren't then we will swap $a[i][j]$ and $a[j][i]$. Otherwise we will keep them the same.
```
Comment (SansPapyrus683) [inline solutions/cf-1713E.mdx:29]:
```suggestion
After this we can construct the answer by iterating over the grid. If it is a diagonal, it will never change. If $a[i][j] == a[j][i]$ then we can manually set them to normal (since we didn't merge them in the disjoint set union when comparing them the states might have not been merged). If it is not a diagonal, we can check whether $i$ and $j$ are merged in the disjoint set union. If they aren't then we will swap $a[i][j]$ and $a[j][i]$. Otherwise, we will keep them the same.
```
what does "iterating over the entire grid"
if it's cell by cell, what does "being a diagonal" mean?

Comment (Keys9873) [reply solutions/cf-1713E.mdx:29]: iterating over grid =
for (int i = 0; i < n; i++) {
    for (int j = 0; j < n; j++) {
        // do stuff to a[i][j]
    }
}

diagonal = a[i][i]

Comment (SansPapyrus683) [reply solutions/cf-1713E.mdx:29]: say this then
"We construct the answer by iterating over each cell of the grid."
"If the cell's row index is equal to its column index (i.e. it's on the diagonal of the grid"

Comment (Keys9873) [reply solutions/cf-1713E.mdx]: fixed

File: solutions/cf-1713E.mdx [hunk 17]
```diff
@@ -0,0 +1,163 @@
...
+				if (dsu.get(i) != dsu.get(j + n)) {
+					dsu.unite(i, j);
+					dsu.unite(i + n, j + n);
+				}
+			}
+		}
+	}
+	
+	// Construct the answer
+	int ans[n][n];
```
Comment (SansPapyrus683) [inline solutions/cf-1713E.mdx:120]: please use a vector

**Review by SansPapyrus683 — CHANGES_REQUESTED** [review]

File: solutions/cf-1713E.mdx (hunk 17 above)
Comment (Keys9873) [reply solutions/cf-1713E.mdx:120]: vectors are too slow

Comment (Keys9873) [reply solutions/cf-1713E.mdx:120]: i got inconsistent results when i used a vector
sometimes tle and sometimes ac

File: solutions/cf-1713E.mdx [hunk 18]
```diff
@@ -0,0 +1,164 @@
...
+				}
+			}
+		}
+	}
+	
+	for (int i = 0; i < n; i++) {
+		for (int j = 0; j < n; j++) {
+			cout << ans[i][j] << " ";
+		}
+		cout << endl;
```
Comment (SansPapyrus683) [inline solutions/cf-1713E.mdx:147]: perhaps try using `\n` here instead of `endl`, see if that fixes the vector issue

**Review by SansPapyrus683 — CHANGES_REQUESTED** [review]

File: solutions/cf-1713E.mdx [hunk 19]
```diff
@@ -0,0 +1,164 @@
...
+author: Aakash Gokhale
+---
+
+## Explanation
+
+[Official Analysis](https://codeforces.com/blog/entry/105419)
+
+Notice that $a[i][j]$ can only be swapped with $a[j][i]$ no matter which operations are performed.
+
+In addition, we want to lexicographically sort the array. Thus, we can iterate and greedily process the array.
```
Comment (SansPapyrus683) [inline solutions/cf-1713E.mdx:14]: > Thus, we can iterate and greedily process the array.

what does this mean

File: solutions/cf-1713E.mdx (hunk 18 above)
Comment (Keys9873) [reply solutions/cf-1713E.mdx]: yea it fixed the issue

File: solutions/cf-1713E.mdx (hunk 19 above)
Comment (Keys9873) [reply solutions/cf-1713E.mdx:14]: Since we want the lexicographically smallest answer, the earlier indices matter more than later indices so we find optimal row swaps for earlier indices first.

Comment (SansPapyrus683) [reply solutions/cf-1713E.mdx:14]: then say that

Comment (Keys9873) [reply solutions/cf-1713E.mdx]: fixed

**Comment by SansPapyrus683** [comment]
@Aadit-Ambadkar review & i'll merge

File: solutions/cf-1713E.mdx [hunk 20]
```diff
@@ -0,0 +1,164 @@
...
+
+Notice that $a[i][j]$ can only be swapped with $a[j][i]$ no matter which operations are performed.
+
+Since we want the lexicographically smallest answer, the earlier indices matter more than later indices so we find optimal row swaps for earlier indices first.
+
+For all $j$ such that $i < j$ (we only need to consider each pair once and keeping $i < j$ will ensure this), we can determine whether we want to swap them or keep them the same.
+
+If $a[i][j] > a[j][i]$, we want to swap them. If $a[i][j] < a[j][i]$ we want to keep them the same. If they are equal it doesn't matter.
+
+However, it is possible that not all conditions can work together with each other and swapping something means we can't swap something else later on.
```
Comment (SansPapyrus683) [inline solutions/cf-1713E.mdx:20]: so...how do we address this problem?

Comment (Keys9873) [reply solutions/cf-1713E.mdx]: fixed

File: solutions/cf-1713E.mdx [hunk 21]
```diff
@@ -0,0 +1,162 @@
...
+
+[Official Analysis](https://codeforces.com/blog/entry/105419)
+
+Notice that $a[i][j]$ can only be swapped with $a[j][i]$ no matter which operations are performed.
+
+For all $j$ such that $i < j$ (we only need to consider each pair once and keeping $i < j$ will ensure this), we can determine whether we want to swap them or keep them the same.
+
+If $a[i][j] > a[j][i]$, we want to swap them. If $a[i][j] < a[j][i]$ we want to keep them the same. If they are equal it doesn't matter.
+
+However, it is possible that not all conditions can work together with each other and swapping something means we can't swap something else later on. Since we want the lexicographically smallest answer, the earlier indices matter more than later indices so we find optimal row swaps for earlier indices first.
```
Comment (SansPapyrus683) [inline solutions/cf-1713E.mdx:18]: > not all conditions can work together with each other

what does this mean exactly?

Comment (Keys9873) [reply solutions/cf-1713E.mdx:18]: swapping/keeping the same $a[i][j]$ with $a[j][i]$ means we can't swap/keep the same $a[k][l]$ with $a[l][k]$ because $i = k$, $i = l$, $j = k$, $j = l$, or some combination.

Does this work as a replacement?

Comment (Keys9873) [reply solutions/cf-1713E.mdx:18]: However, it is possible that not all conditions can work together with each other and swapping something/keeping something the same means we can't swap something else/keep something the same late on. For example it is impossible to swap $i$ and $j$, keep $j$ and $k$ the same, and keep $i$ and $k$ the same. We can first flip the $i$ith column to swap $i$ and $j$. We do nothing to $j$ or $k$ since we want to keep them the same. In order to keep $i$ and $k$ the same, we need to flip the $k$th column because we earlier swapped $i$ and two swaps will cancel out. However, swapping $k$ will mean that the second condition will fail and we have reached a scenario where we cannot satisfy all conditions.

Comment (Keys9873) [reply solutions/cf-1713E.mdx:18]: is this better?

Comment (SansPapyrus683) [reply solutions/cf-1713E.mdx:18]: at this point i'd just give an example

Comment (Keys9873) [reply solutions/cf-1713E.mdx:18]: For example it is impossible to swap $i$ and $j$, keep $j$ and $k$ the same, and keep $i$ and $k$ the same.

Isn't this already an example?

Comment (SansPapyrus683) [reply solutions/cf-1713E.mdx:18]: > swap $i$ and $j$

how can you swap $i$ and $j$? are these numbers? points?
instead of variables, just use concrete numbers, perhaps with the sample test case

Comment (Keys9873) [reply solutions/cf-1713E.mdx:18]: However, it is possible that not all conditions can work together with each other and swapping something/keeping something the same means we can't swap something else/keep something the same late on. For example it is impossible to swap a[i][j] and a[j][i], keep a[j][k] and a[k][j] the same, and keep a[i][k] and a[k][i] the same. We can first flip the ith column/row which satisfies the first condition. We do nothing to the jth or kth column/row since we want to keep them the same which satisfies the second condition. In order to keep a[i][k] and a[k][i] the same, we need to flip the kth column/row and two swaps (i and k) will cancel out. However, swapping the kth column/row will cause the second condition to now fail and we have reached a scenario where we cannot satisfy all conditions.

Comment (Keys9873) [reply solutions/cf-1713E.mdx:18]: I was unclear in my explanation. I was supposed to reference a[i][j], a[j][k], and a[i][k]

Comment (SansPapyrus683) [reply solutions/cf-1713E.mdx:18]: this is pretty hard to follow
i'd use the example input or make an example array & say "it's possible to swap 3 & 4, but then this would [something bad]"

Comment (Keys9873) [reply solutions/cf-1713E.mdx:18]: However, it is possible that not all conditions can work together with each other and swapping something/keeping something the same means we can't swap something else/keep something the same later on.
Consider to following input:
0 2 1
1 0 1
3 2 0
We want to first swap a[0][1] and a[1][0] because 1 is less than 2. We can do this by flipping the 0th row/col. This would give us the array:
0 1 3
2 0 1
1 2 0
Then we want to swap a[0][2] and a[2][0] because 1 is less than 3. To do this either we can either swap the 0th row/col or the 2nd row/col. However, swapping the 0th row/col would undo the efforts of our first operation so we will swap the 2nd row/col and get:
0 1 1
2 0 2
3 1 0
Now we want to swap a[1][2] and a[2][1] because 1 is less than 2 and we want the lexicographically minimum. However swapping the 1st row/col would mess up the swap we made in the first operation and swapping the 2nd row/col would mess up the swap we made in the second operation.
Thus, it is impossible to make all swaps optimally.

Is this better?

Comment (SansPapyrus683) [reply solutions/cf-1713E.mdx:18]: much much better

Comment (Keys9873) [reply solutions/cf-1713E.mdx]: fixed

File: solutions/cf-1713E.mdx [hunk 22]
```diff
@@ -0,0 +1,184 @@
...
+For all $j$ such that $i < j$ (we only need to consider each pair once and keeping $i < j$ will ensure this), we can determine whether we want to swap them or keep them the same.
+
+If $a[i][j] > a[j][i]$, we want to swap them. If $a[i][j] < a[j][i]$ we want to keep them the same. If they are equal it doesn't matter.
+
+However, it is possible that not all conditions can work together with each other and swapping something/keeping something the same means we can't swap something else/keep something the same later on.
+Consider to following input:
+
+$0$ $2$ $1$
+$1$ $0$ $1$
+$3$ $2$ $0$
```
Comment (SansPapyrus683) [inline solutions/cf-1713E.mdx:23]: did you check the formatting on the website? this seems kinda sus

**Comment by SansPapyrus683** [comment]
hold on where's the problem in the guide itself? it doesn't appear when i search it

File: solutions/cf-1713E.mdx (hunk 22 above)
Comment (Keys9873) [reply solutions/cf-1713E.mdx]: fixed

**Comment by SansPapyrus683** [comment]
strange, this problem's in a modulel but it doesn't appear in the search thing?

**Comment by Keys9873** [comment]
I don't know why it isn't in the search thing. I added it to the extraPractice json file. Do I need to add it somewhere else?

**Comment by SansPapyrus683** [comment]
idk, good enough i guess

**Comment by SansPapyrus683** [comment]
image
typo

**Comment by Keys9873** [comment]
Fixed
Istg I am so done with this PR

**Comment by SansPapyrus683** [comment]
same, bro, same
